lib/thesis/notifications.ex

```diff
@@ -0,0 +1,163 @@
+defmodule Thesis.Notifications do
+  @moduledoc """
+  Builds a map of Notifications that is passed to the client side. The key should be a
+  representation of the category/type of notification (this is used to determine where and how it's
+  displayed). The value is a list of notification strings.
+
+  Example of final map:
+  %{
+    "page-settings": [],
+    "add-page": [],
+    "import-export-restore": [],
+    "other": []
+  }
+  """
+
+  alias Thesis.Backup
+  import Thesis.Config
+  import Thesis.Utilities
+  import Ecto.Query, only: [limit: 2]
+
+  @allowed_notification_keys [
+    "page-settings",
+    "add-page",
+    "import-export-restore",
+    "other",
+    :page_settings,
+    :add_page,
+    :import_export_restore,
+    :other
+  ]
+
+  def all(page) do
+    %{
+      "page-settings" => [],
+      "add-page" => [],
+      "import-export-restore" => [],
+      "other" => []
+    }
+    |> notifications_from_host_app(notifications())
+    |> notifications_regarding_env(Mix.env)
+    |> notifications_regarding_page(page)
+  end
+
+  def notifications_regarding_page(acc, nil), do: acc
+  def notifications_regarding_page(acc, %_{}) do
+    acc
+  end
+
+  def notifications_regarding_env(acc, :test), do: acc
+  def notifications_regarding_env(acc, _) do
+    acc
+    |> notification_to_run_migration_for_version_0_2_1()
+  end
+
+  @doc """
+  Adds notifications provided by the host app to the accumulator.
+
+      iex> notifications_from_host_app(%{"page-settings" => []}, [page_settings: ["notif"]])
+      %{"page-settings" => ["notif"]}
+      iex> notifications_from_host_app(%{"page-settings" => []}, [page_settings: :invalid])
+      %{"page-settings" => []}
+      iex> notifications_from_host_app(%{"page-settings" => ["notif1"]}, [page_settings: ["notif2"]])
+      %{"page-settings" => ["notif2", "notif1"]}
+      iex> notifications_from_host_app(%{"other" => [], "add-page" => ["notif"]}, [other: ["notif"]])
+      %{"other" => ["notif"], "add-page" => ["notif"]}
+      iex> notifications_from_host_app(%{"other" => []}, nil)
+      %{"other" => []}
+  """
+  def notifications_from_host_app(acc, notifications) when is_list(notifications) do
+    case verify_notifications_structure(notifications) do
+      true ->
+        Enum.reduce(notifications, acc, fn({type, list}, acc) ->
+          update_notifications_map(acc, type, list)
+        end)
+      _ -> acc
+    end
+  end
+  def notifications_from_host_app(acc, _), do: acc
```

Comment:
Here, we accumulate notifications from the host app. First we verify that the provided configs are in the correct format, then we add them to the accumulator.